kds/kds.go

```diff
@@ -227,33 +323,104 @@ func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+		tcbVersion.version = TCBStructVersion1
+		tcbVersion.TCB = (uint64(parts.UcodeSpl) << 56) |
+			(uint64(parts.SnpSpl) << 24) |
+			(uint64(parts.TeeSpl) << 16) |
+			(uint64(parts.BlSpl) << 8) |
+			(uint64(parts.FmcSpl) << 0)
+	default:
+		return tcbVersion, fmt.Errorf("unsupported TCB struct version: %v", parts.version)
+	}
+
+	return tcbVersion, nil
+}
+
+func (parts TCBParts) String() string {
+	switch parts.version {
+	case TCBStructVersion1:
+		return fmt.Sprintf("{BlSpl:%d TeeSpl:%d Spl5:%d Spl6:%d Spl7:%d SnpSpl:%d UcodeSpl:%d FmcSpl: %d}",
+			parts.BlSpl, parts.TeeSpl, parts.Spl5, parts.Spl6, parts.Spl7, parts.SnpSpl, parts.UcodeSpl, parts.FmcSpl)
+	default:
+		return fmt.Sprintf("{BlSpl:%d TeeSpl:%d Spl4:%d Spl5:%d Spl6:%d Spl7:%d SnpSpl:%d UcodeSpl:%d}",
+			parts.BlSpl, parts.TeeSpl, parts.Spl4, parts.Spl5, parts.Spl6, parts.Spl7, parts.SnpSpl, parts.UcodeSpl)
+	}
+}
+
+// ComposeTCBParts returns an SEV-SNP TCB_VERSION from OID mapping values. The spl4-spl7 fields are
+// reserved, but the KDS specification designates them as 4 byte-sized fields.
+//
+// Deprecated: use ComposeTCBPartsToTCBVersionStruct instead
+func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+	tcb, err := parts.ToTCBVersionStruct()
+	if err != nil {
+		return TCBVersion(0), err
+	}
+
+	return TCBVersion(tcb.TCB), nil
+}
+
+// ComposeTCBPartsToTCBVersionStruct returns an SEV-SNP TCB_VERSION from OID mapping
+// values as a TCBVersionStruct. The spl4-spl7 fields are reserved, but the
+// KDS specification designates them as 4 byte-sized fields.
+func ComposeTCBPartsToTCBVersionStruct(parts TCBParts) (TCBVersionStruct, error) {
+	return parts.ToTCBVersionStruct()
 }
 
 // DecomposeTCBVersion interprets the byte components of the AMD representation of the
 // platform security patch levels into a struct.
+//
+// Deprecated: use DecomposeTCBVersionStruct instead
 func DecomposeTCBVersion(tcb TCBVersion) TCBParts {
-	return TCBParts{
-		UcodeSpl: uint8((uint64(tcb) >> 56) & 0xff),
-		SnpSpl:   uint8((uint64(tcb) >> 48) & 0xff),
-		Spl7:     uint8((uint64(tcb) >> 40) & 0xff),
-		Spl6:     uint8((uint64(tcb) >> 32) & 0xff),
-		Spl5:     uint8((uint64(tcb) >> 24) & 0xff),
-		Spl4:     uint8((uint64(tcb) >> 16) & 0xff),
-		TeeSpl:   uint8((uint64(tcb) >> 8) & 0xff),
-		BlSpl:    uint8((uint64(tcb) >> 0) & 0xff),
+	parts, _ := DecomposeTCBVersionStruct(tcb)
+	return parts
+}
+
+// DecomposeTCBVersionStruct accepts SEV-SNP TCB version as TCBVersion,
+// TCBVersionStruct or uint64 and  decodes various security patch
+// levels from them.
+func DecomposeTCBVersionStruct(t any) (TCBParts, error) {
+	var tcb TCBVersionStruct
+
+	switch v := t.(type) {
+	case TCBVersionStruct:
+		tcb = v
+	case *TCBVersionStruct:
+		if v == nil {
+			return TCBParts{}, fmt.Errorf("TCBVersionStruct is nil")
+		}
+		tcb = *v
+	case TCBVersion:
+		tcb.FromTCBVersion(v)
+	case *TCBVersion:
+		if v == nil {
+			return TCBParts{}, fmt.Errorf("TCBVersion is nil")
+		}
```

Comment:
can't the nil comparison be lifted out of the switch?

Reply:
ACK